kratos/utilities/assign_master_slave_constraints_to_neighbours_utility.cpp

```diff
@@ -0,0 +1,247 @@
+    } else {
+        std::stringstream variables_str;
+        variables_str << "[";
+        for (const auto& variable : rVariableList) {
+            variables_str << variable.get() << ", ";
+        }
+        variables_str.seekp(-2, std::ios_base::end); // Remove the last ", "
+        variables_str << "]";
+        KRATOS_ERROR << "The node with ID " << pSlaveNode->Id() << " does not have the required DOFs for one of the variables " << variables_str.str() << std::endl;
+    }
+
+    rSlaveCoordinates = pSlaveNode->Coordinates();
+
+    KRATOS_CATCH("");
+}
+
+// Get DOFs and Coordinates arrays for a given variable list for a cloud of nodes
+void AssignMasterSlaveConstraintsToNeighboursUtility::GetDofsAndCoordinatesForCloudOfNodes(
+    const ResultNodesContainerType& CloudOfNodesArray,
+    const std::vector<std::reference_wrapper<const Kratos::Variable<double>>>& rVariableList,
+    std::vector<DofPointerVectorType>& rCloudOfDofs,
+    Matrix& rCloudOfNodesCoordinates
+)
+{
+    KRATOS_TRY;
+
+    // Resize rCloudOfDofs to match the size of the variable list
+    rCloudOfDofs.resize(rVariableList.size());
+
+    // Resize rCloudOfDofs vectors to match the size of the cloud of nodes
+    for (auto& dofs : rCloudOfDofs)
+    {
+        dofs.resize(CloudOfNodesArray.size());
+    }
+
+    // Resize rCloudOfNodesCoordinates to match the size of the cloud of nodes
+    rCloudOfNodesCoordinates.resize(CloudOfNodesArray.size(), 3);
+
+    // Loop over the cloud of nodes
+    for (int i = 0; i < static_cast<int>(CloudOfNodesArray.size()); i++)
+    {
+        NodeType::Pointer pSlaveNode = CloudOfNodesArray[i];
+
+        // Loop over the variable list
+        for (int j = 0; j < static_cast<int>(rVariableList.size()); j++)
+        {
+            const auto& rVariable = rVariableList[j];
+
+            // Check if the slave node has the required DOF for the current variable
+            if (pSlaveNode->HasDofFor(rVariable.get()))
+            {
+                // Assign the DOF pointer to the corresponding position in rCloudOfDofs
+                rCloudOfDofs[j][i] = pSlaveNode->pGetDof(rVariable.get());
+            }
+            else
+            {
+                std::stringstream variable_str;
+                variable_str << rVariable.get();
+                KRATOS_ERROR << "The node with ID " << pSlaveNode->Id() << " does not have the required DOF for the variable " << variable_str.str() << std::endl;
+            }
+        }
+
+        // Assign the coordinates of the slave node to the corresponding row in rCloudOfNodesCoordinates
+        noalias(row(rCloudOfNodesCoordinates, i)) = pSlaveNode->Coordinates();
+    }
+
+    KRATOS_CATCH("");
+}
+
+void AssignMasterSlaveConstraintsToNeighboursUtility::AssignMasterSlaveConstraintsToNodes(
+    NodesContainerType pSlaveNodes,
+    double const Radius,
+    ModelPart& rComputingModelPart,
+    const std::vector<std::reference_wrapper<const Kratos::Variable<double>>>& rVariableList,
```

Comment:
I didn't know about std::reference_wrapper, until now we were using variables raw pointers...